README.md

```diff
@@ -174,156 +169,380 @@ fetch('http://example.com/posts', {
-  json?: unknown
-  /** Object that can be passed to `serialize` */
-  params?: Record<string, any>
-  /** Throw `TimeoutError` if timeout is passed */
-  timeout?: number
-  /** String that will prepended to `resource` in `fetch` instance */
-  prefixUrl?: string
-  /** Request headers */
-  headers?: Record<string, string>
-  /**`node-fetch` v3 option, default is 10mb */
-  highWaterMark?: number
-  /** Request headers, can be async */
-  getHeaders?(resource: string, init: RequestInit): Headers | Promise<Headers>
-  /** Custom params serializer, default to `URLSearchParams` */
-  serialize?(params: Record<string, any>): URLSearchParams | string
-  /** Response handler, must handle status codes or throw `ResponseError` */
-  onResponse?(
-    response: Response,
-    options: Options
-  ): Response | Promise<Response>
-  /** Response handler with sucess status codes 200-299 */
-  onSuccess?(response: Response, options: Options): Response | Promise<Response>
-  /** Error handler. Throw passed `error` for unhandled cases, throw custom errors, or return the new `Response` */
-  onFailure?(
-    error: ResponseError | AbortError | TimeoutError | TypeError | Error,
-    options: Options
-  ): Response | Promise<Response>
-  /** Transform parsed JSON from response */
-  onJSON?(input: unknown): unknown
+import queryString from 'query-string'
+
+const instance = YF.create({
+  resource: 'https://jsonplaceholder.typicode.com',
+  serialize: (params) =>
+    queryString.stringify(params, {
+      arrayFormat: 'bracket',
+    }),
+})
+
+// request will be sent to 'https://jsonplaceholder.typicode.com/posts?userId=1&tags[]=1&tags[]=2'
+await instance.get('/posts', { params: { userId: 1, tags: [1, 2] } })
+```
+
+#### timeout?: number
+
+If specified `TimeoutError` will be thrown and the request will be cancelled after the specified duration.
+
+```ts
+try {
+  await instance.get('/posts', { timeout: 500 })
+} catch (error) {
+  if (error instanceof TimeoutError) {
+    // do something, or nothing
+  }
 }
 ```
 
-### ResponseBody
+#### onRequest?(options: RequestOptions): Promise\<void> | void
+
+Request handler. Use the callback to modify options before the request or cancel the request. Please, note the options here are in the final state before the request will be made, this means `resource` is string, `params` is instance of `URLSearchParams`, and `headers` is instance of `Headers`.
 
 ```ts
-interface ResponseBody extends Promise<Response> {
-  json<T>(): Promise<T>
-  text(): Promise<string>
-  blob(): Promise<Blob>
-  arrayBuffer(): Promise<ArrayBuffer>
-  formData(): Promise<FormData>
-  void(): Promise<void>
+const authorized = instance.extend({
+  async onRequest(options) {
+    // fetch 'token' somehow
+    options.headers.set('Authorization', `Bearer ${await 'token'}`)
+  },
+})
+
+// will include `Authorization` will be resolved with async `Bearer token`.
+await authorized.get('/posts')
+```
+
+```ts
+const cancellable = instance.extend({
```

Comment:
it could be nice to have an "abort controller support" in the top list linking to this example, wdyt?
(and maybe this example might be moved to options list into signal header section)